### PR TITLE
Support for China‘s TLS handshake protocol (GM/T 0024-2014).

### PR DESCRIPTION
**Pull request by jntass**
Although China algorithms (e.g. SM2,SM3,SM4) are included, the China TLS handshake protocol is not. So our patch is to implement the China TLS handshake protocol according GM/T 0024-2014 standard. All the patch can be ignored by the macro definition of “#define OPENSSL_NO_CNSM".

We also open the source of Tassl(https://github.com/jntass/TASSL-1.1.1b) on github based on openssl-1.1.1b. For more detail about us, please view our website http://www.tass.com.cn.

##### Checklist

- [ ] documentation is added or updated
- [x] tests are added or updated
CLA: trivial

### CODE REVIEW

**Comment by mspncp** [comment]
Welcome @jntass and thank you for your contribution! Please note that contributions need to be covered by a Contributor License Agreement (CLA), see https://www.openssl.org/policies/cla for the details. In your case, I think that both an Individual CLA (ICLA) and a Company CLA (CCLA) will be required.

**Comment by mspncp** [comment]
Also, please not that for new features we only accept pull requests against the `master` branch, so you will have to rebase it.

**Comment by jntass** [comment]
Thank you for quickly response! About CLA, we'll finish it in the next week. About changing the base, it'll be done next week also.

**Comment by mspncp** [comment]
Please also have a look at the continuous-integration builds (Travis, AppVeyor). They currently fail because the OpenSSL testsuite fails. For example:

```
../../_srcdist/test/recipes/70-test_recordlen.t ................ 

1..1

    # Subtest: ../../test/recordlentest

    1..1

        # Subtest: test_record_overflow

        1..6

        ok 1 - iteration 1

        ok 2 - iteration 2

        ok 3 - iteration 3

        ok 4 - iteration 4

        # INFO:  @ ../_srcdist/test/ssltestlib.c:768

        # SSL_accept() failed -1, 1

        # INFO:  @ ../_srcdist/test/ssltestlib.c:752

        # SSL_connect() failed -1, 1

        # ERROR: (bool) 'create_ssl_connection(serverssl, clientssl, SSL_ERROR_NONE) == true' failed @ ../_srcdist/test/recordlentest.c:150

        # false

        # 140735837094720:error:14094438:SSL routines:ssl3_read_bytes:tlsv1 alert internal error:../_srcdist/ssl/record/rec_layer_s3.c:1536:SSL alert number 80

        not ok 5 - iteration 5

        # INFO:  @ ../_srcdist/test/ssltestlib.c:768

        # SSL_accept() failed -1, 1

        # INFO:  @ ../_srcdist/test/ssltestlib.c:752

        # SSL_connect() failed -1, 1

        # ERROR: (bool) 'create_ssl_connection(serverssl, clientssl, SSL_ERROR_NONE) == true' failed @ ../_srcdist/test/recordlentest.c:150

        # false

        # 140735837094720:error:14094438:SSL routines:ssl3_read_bytes:tlsv1 alert internal error:../_srcdist/ssl/record/rec_layer_s3.c:1536:SSL alert number 80

        not ok 6 - iteration 6

    not ok 1 - test_record_overflow
```

(see https://api.travis-ci.org/v3/job/530677739/log.txt)

File: apps/s_client.c [hunk 1]
```diff
@@ -2590,7 +2638,7 @@ int s_client_main(int argc, char **argv)
             do {
                 mbuf_len = BIO_gets(fbio, mbuf, BUFSIZZ);
                 /*
-                 * According to RFC 5804 § 1.7, capability
+                 * According to RFC 5804 搂 1.7, capability
```
Comment (kroeckx) [inline apps/s_client.c]: Please don't change that to Chinese. Maybe write it as "section" instead?

File: ssl/statem/statem_srvr.c [hunk 2]
```diff
@@ -774,10 +794,6 @@ static ossl_inline int conn_is_closed(void)
 #if defined(ECONNRESET)
     case ECONNRESET:
         return 1;
-#endif
-#if defined(WSAECONNRESET)
-    case WSAECONNRESET:
-        return 1;
```
Comment (kroeckx) [inline ssl/statem/statem_srvr.c]: Why is this removed?

File: ssl/statem/statem_srvr.c [hunk 3]
```diff
@@ -2448,6 +2464,9 @@ int tls_construct_server_hello(SSL *s, WPACKET *pkt)
             return 0;
         }
     } else if (!(s->verify_mode & SSL_VERIFY_PEER)
+    #ifndef OPENSSL_NO_CNSM
+	&&(s->s3 && s->s3->tmp.new_cipher && s->s3->tmp.new_cipher->id != TLS1_CK_ECDHE_WITH_SM4_SM3)  //ECDHE-SM4-SM3 will use the s->s3->hansdshake_buff later,so shouldn't digest the cached record here modify by gujq on 20190313
```
Comment (kroeckx) [inline ssl/statem/statem_srvr.c]: That comment should be on it's own line, and use /* and */ for comment.

File: ssl/statem/statem_srvr.c [hunk 4]
```diff
@@ -2968,6 +3070,161 @@ static int tls_process_cke_psk_preamble(SSL *s, PACKET *pkt)
...
+     }
+     
+     
+    if (decrypt_len < 0) {
+        SSLfatal(s, SSL_AD_DECRYPT_ERROR, SSL_F_TLS_PROCESS_CKE_SM2ECC,
+                 ERR_R_INTERNAL_ERROR);
+        goto err;
+    }
+			
+		#ifdef CIPHER_DEBUG
```
Comment (kroeckx) [inline ssl/statem/statem_srvr.c]: Don't use tabs, use spaces.

File: ssl/statem/statem_srvr.c [hunk 5]
```diff
@@ -2968,6 +3070,161 @@ static int tls_process_cke_psk_preamble(SSL *s, PACKET *pkt)
...
+     * generates a random premaster secret for the case that the decrypt
+     * fails. See https://tools.ietf.org/html/rfc5246#section-7.4.7.1
+     */
+
+    if (RAND_priv_bytes(rand_premaster_secret,
+                      sizeof(rand_premaster_secret)) <= 0) {
+        SSLfatal(s, SSL_AD_INTERNAL_ERROR, SSL_F_TLS_PROCESS_CKE_SM2ECC,
+                 ERR_R_INTERNAL_ERROR);
+        goto err;
+    }
```
Comment (kroeckx) [inline ssl/statem/statem_srvr.c]: This doesn't actually seem to be properly used, you just decrypt to that buffer.

File: ssl/statem/statem_srvr.c [hunk 6]
```diff
@@ -2968,6 +3070,161 @@ static int tls_process_cke_psk_preamble(SSL *s, PACKET *pkt)
...
+    	    goto err;
+    	}
+
+     /* Need to call EVP_PKEY_CTX_CTL(pkey_ctx, EVP_PKEY_EC, EVP_PKEY_CTRL_MD, 0, (void *)EVP_sm3()) */
+     decrypt_len = sizeof(rand_premaster_secret);
+     if (EVP_PKEY_decrypt(pkey_ctx, rand_premaster_secret, (size_t *)&decrypt_len, PACKET_data(&enc_premaster), (int)PACKET_remaining(&enc_premaster)) <= 0)
+     {
+         SSLfatal(s, SSL_AD_DECODE_ERROR, SSL_F_TLS_PROCESS_CKE_SM2ECC,
+                     ERR_R_INTERNAL_ERROR);
+         goto err;
```
Comment (kroeckx) [inline ssl/statem/statem_srvr.c]: This is just wrong.

File: include/openssl/tls1.h [hunk 7]
```diff
@@ -24,6 +24,9 @@ extern "C" {
 #  define OPENSSL_TLS_SECURITY_LEVEL 1
 # endif
 
+#ifndef OPENSSL_NO_CNSM
+# define SM1_1_VERSION                   0x0101
+#endif
```
Comment (kroeckx) [inline include/openssl/tls1.h]: So this is a different protocol than SSL/TLS, looks a lot like TLS, but uses a different version number?
@richsalz or @kaduk, do you care to comment on this?

File: ssl/record/ssl3_record.c [hunk 8]
```diff
@@ -309,7 +309,11 @@ int ssl3_get_record(SSL *s)
                     return -1;
                 }
 
-                if ((version >> 8) != SSL3_VERSION_MAJOR) {
+                #ifndef OPENSSL_NO_CNSM
+                    if ((version >> 8) != SSL3_VERSION_MAJOR && (version != SM1_1_VERSION)) {
```
Comment (richsalz) [inline ssl/record/ssl3_record.c]: not sure why both parts of the ```&&``` are needed.

File: include/openssl/tls1.h (hunk 7 above)
Comment (richsalz) [reply include/openssl/tls1.h]: This is not TLS.  It is a national variant. Does OpenSSL want to support this?
It uses many values that the IETF consider "its own" as defined in the registries at https://www.iana.org/assignments/tls-parameters/tls-parameters.xhtml  The handshake/version number here, but also the cipher assignments used elsewhere in the code.

My recommendation has two parts. First, someone (perhaps someone at TASS?) should apply to the IANA registry and get assigned numbers. Once that is done, the project should decide if it wants to implement this. 

As it stands now, if you implement this, you could run into interoperability problems if/when IETF TLS assigns conflicting curve, etc., values, and I recommend rejecting this PR.

Hope this info helps you decide.

Comment (kroeckx) [reply include/openssl/tls1.h]: So one of the things about this PR is that they add a CNTLS_client_method(), but don't add a CNTLS_server_method(), and I think you just need to use the TLS_server_method(), and I think the TLS_server_method() they just accepts some other protocol.

I think they should either fully use a different protocol with different methods (like TLS vs DTLS), or actually use TLS and add the needed extensions to that.

File: apps/s_client.c (hunk 1 above)
Comment (jntass) [reply apps/s_client.c]: ok，will change it to section.

File: ssl/statem/statem_srvr.c (hunk 5 above)
Comment (jntass) [reply ssl/statem/statem_srvr.c]: Yes, it's really not necessary. I just copy the part implement of tls_process_cke_rsa() , and don't check the purpose of those code carefully.

File: ssl/statem/statem_srvr.c (hunk 6 above)
Comment (jntass) [reply ssl/statem/statem_srvr.c]: What you mean is  about the not necessary code  above? or the line 3117 to 3131 is wrong ?

File: ssl/statem/statem_srvr.c (hunk 4 above)
Comment (jntass) [reply ssl/statem/statem_srvr.c]: got it.

File: include/openssl/tls1.h (hunk 7 above)
Comment (jntass) [reply include/openssl/tls1.h]: Yes, indeed the CNSM is really a national variant. It follow the main of tls1.1, but it has lots of different from the official tls. Such as the version num, use double certificate when handshake (When encryption is required, use the encryption certificate. When authentication is required, use the authentication certificate.)  We have the CNSM internally, so we develop the tassl for use by ourself. After we use, we also want to contribute the tassl to the official for those needed especially we see the  cryptographic algorithm of 'smx' series are released by the official openssl.
As for the follow-up, we are waiting for your decision and see our plan. Thank all of you! @kroeckx @richsalz @mspncp

File: ssl/statem/statem_srvr.c (hunk 4 above)
Comment (mspncp) [reply ssl/statem/statem_srvr.c]: @jntass to avoid change requests related to coding style issues, it might be helpful to take a look at our [Coding Style Guidelines](https://www.openssl.org/policies/codingstyle.html) on our website.

File: include/openssl/tls1.h (hunk 7 above)
Comment (kroeckx) [reply include/openssl/tls1.h]: Since the IETF plans to deprecate TLS 1.0 and 1.1 (https://tools.ietf.org/html/draft-ietf-tls-oldversions-deprecate), and this at least seems to have some of the same issues, I don't think we should add it.

File: Configurations/unix-Makefile.tmpl [hunk 9]
```diff
@@ -506,6 +506,7 @@ install_dev: install_runtime_libs
 	@chmod 644 $(DESTDIR)$(INSTALLTOP)/include/openssl/applink.c
 	@ : {- output_on() unless grep { $_ eq "OPENSSL_USE_APPLINK" } (@{$target{defines}}, @{$config{defines}}); "" -}
 	@set -e; for i in $(SRCDIR)/include/openssl/*.h \
+			  $(SRCDIR)/crypto/include/internal/*.h \
```
Comment (levitte) [inline Configurations/unix-Makefile.tmpl]: Absolutely not. Internal headers should *not* get publically installed, that's asking for trouble.

File: crypto/ec/ec_pmeth.c [hunk 10]
```diff
@@ -323,7 +392,10 @@ static int pkey_ec_ctrl(EVP_PKEY_CTX *ctx, int type, int p1, void *p2)
             EVP_MD_type((const EVP_MD *)p2) != NID_sha224 &&
             EVP_MD_type((const EVP_MD *)p2) != NID_sha256 &&
             EVP_MD_type((const EVP_MD *)p2) != NID_sha384 &&
-            EVP_MD_type((const EVP_MD *)p2) != NID_sha512) {
+            EVP_MD_type((const EVP_MD *)p2) != NID_sha512 &&
+            #ifndef OPENSSL_NO_CNSM
+            EVP_MD_type((const EVP_MD *)p2) != NID_sm3) {
+            #endif
```
Comment (levitte) [inline crypto/ec/ec_pmeth.c]: The ending `) {` should be outside the guard. The way it looks now, build will fail when configured `no-cnsm`

Comment (levitte) [reply crypto/ec/ec_pmeth.c]: (by the way, you need to add the string `"cnsm"` to the `@disablables` array in `Configure`)

File: crypto/sm2/sm2_crypt.c [hunk 11]
```diff
@@ -391,3 +391,350 @@ int sm2_decrypt(const EC_KEY *key,
 
     return rc;
 }
+
+#ifndef OPENSSL_NO_CNSM
+/* GM/T003_2012 Defined Key Derive Function */
+int kdf_gmt003_2012(unsigned char *out, size_t outlen, const unsigned char *Z, size_t Zlen, const unsigned char *SharedInfo, size_t SharedInfolen, const EVP_MD *md)
```
Comment (levitte) [inline crypto/sm2/sm2_crypt.c]: Is this meant to be an internal function? Then it should be made `static`. Otherwise, it should probably be implemented in terms of the new `EVP_KDF` API.

File: include/openssl/tls1.h [hunk 12]
```diff
@@ -1127,6 +1147,22 @@ __owur int SSL_check_chain(SSL *s, X509 *x, EVP_PKEY *pk, STACK_OF(X509) *chain)
 # define TLS1_TXT_RSA_PSK_WITH_ARIA_128_GCM_SHA256         "RSA-PSK-ARIA128-GCM-SHA256"
 # define TLS1_TXT_RSA_PSK_WITH_ARIA_256_GCM_SHA384         "RSA-PSK-ARIA256-GCM-SHA384"
 
+# ifndef OPENSSL_NO_CNSM
+#  define TLS1_TXT_ECDHE_WITH_SM1_SM3                    "ECDHE-SM1-SM3"
+#  define TLS1_TXT_ECC_WITH_SM1_SM3                      "ECC-SM1-SM3"
+#  define TLS1_TXT_IBSDH_WITH_SM1_SM3                    "IBSDH-SM1-SM3"
+#  define TLS1_TXT_IBC_WITH_SM1_SM3                      "IBC-SM1-SM3"
+#  define TLS1_TXT_RSA_WITH_SM1_SM3                      "RSA-SM1-SM3"
+#  define TLS1_TXT_RSA_WITH_SM1_SHA1                     "RSA-SM1-SHA1"
```
Comment (levitte) [inline include/openssl/tls1.h]: What is sm1?

File: ssl/ssl_rsa.c [hunk 13]
```diff
@@ -334,6 +343,94 @@ static int ssl_set_cert(CERT *c, X509 *x)
...
+        if (x && (X509_get_extension_flags(x) & EXFLAG_KUSAGE) && (X509_get_key_usage(x)& X509v3_KU_DIGITAL_SIGNATURE)){
+             if (c->pkeys[i].privatekey != NULL) {
+	        /*
+	         * The return code from EVP_PKEY_copy_parameters is deliberately
+	         * ignored. Some EVP_PKEY types cannot do this.
+	         */
+	        EVP_PKEY_copy_parameters(pkey, c->pkeys[i].privatekey);
+	        ERR_clear_error();
+
+		#ifndef OPENSSL_NO_RSA
```
Comment (levitte) [inline ssl/ssl_rsa.c]: Indentation of preprocessor directives in this way is not ok in our code.

File: ssl/statem/statem_lib.c [hunk 14]
```diff
@@ -995,6 +1063,133 @@ static int ssl_add_cert_chain(SSL *s, WPACKET *pkt, CERT_PKEY *cpk)
...
+         * the cert - we're just building as much of the chain as we can
+         */
+        (void)X509_verify_cert(xs_ctx);
+        /* Don't leave errors in the queue */
+        ERR_clear_error();
+        chain = X509_STORE_CTX_get0_chain(xs_ctx);
+        i = ssl_security_cert_chain(s, chain, NULL, 0);
+        if (i != 1) {
+#if 0
+            /* Dummy error calls so mkerr generates them */
```
Comment (levitte) [inline ssl/statem/statem_lib.c]: This should not be necessary

File: ssl/t1_lib.c [hunk 15]
```diff
@@ -181,22 +184,56 @@ static const uint16_t eccurves_default[] = {
...
     if (group_id < 1 || group_id > OSSL_NELEM(nid_list))
         return NULL;
     return &nid_list[group_id - 1];
 }
 
+#ifndef OPENSSL_CNSM
+uint16_t tls1_nid2group_id(int nid)
+#else
 static uint16_t tls1_nid2group_id(int nid)
+#endif
```
Comment (levitte) [inline ssl/t1_lib.c]: I have seen this construct in several places, and find it unnecessary. If you need to remove `static` from a function for some reason, that's OK. But please do double check that it's actually necessary.

File: Configurations/unix-Makefile.tmpl (hunk 9 above)
Comment (jntass) [reply Configurations/unix-Makefile.tmpl]: It's a mistake. What i really want to copy is /include/openssl/sm*.h instead  of those internal header file.  Because i found when after "make install" finished, in the install directory, there is only  sm2_sign( ) in libcrypto.a , but no definition header file in "include/openssl". If i want to invoke the function of sm2_sign, i don't have a header file to included. So i just did it  as RSA_sign define in "include/openssl/rsa.h" did,  copy the rsa.h to "include/openssl" directory".
Correct me if i misunderstood the usage!

File: crypto/ec/ecdsa_sign.c [hunk 16]
```diff
@@ -29,7 +35,13 @@ ECDSA_SIG *ECDSA_do_sign_ex(const unsigned char *dgst, int dlen,
 int ECDSA_sign(int type, const unsigned char *dgst, int dlen, unsigned char
                *sig, unsigned int *siglen, EC_KEY *eckey)
 {
-    return ECDSA_sign_ex(type, dgst, dlen, sig, siglen, NULL, NULL, eckey);
+#ifndef OPENSSL_NO_CNSM
+
+    if (EC_GROUP_get_curve_name(EC_KEY_get0_group(eckey)) == NID_sm2)
+        return  sm2_sign(dgst, dlen, sig, siglen, eckey);
+     else
```
Comment (paulidale) [inline crypto/ec/ecdsa_sign.c]: The `else` isn't required because of the return on the previous line (which has a double space after it).

File: apps/s_client.c [hunk 17]
```diff
@@ -619,10 +619,16 @@ const OPTIONS s_client_options[] = {
 #endif
     {"verify", OPT_VERIFY, 'p', "Turn on peer certificate verification"},
     {"cert", OPT_CERT, '<', "Certificate file to use, PEM format assumed"},
+#ifndef OPENSSL_NO_CNSM
+    {"cert_enc", OPT_CERT_ENC, '<', "Encrypto Certificate file to use when use cnsm, PEM format assumed"},
```
Comment (paulidale) [inline apps/s_client.c]: Did you mean `Encryption` instead of `Encrypto` here and below?

File: apps/s_client.c [hunk 18]
```diff
@@ -1913,6 +1942,25 @@ int s_client_main(int argc, char **argv)
     if (!set_cert_key_stuff(ctx, cert, key, chain, build_chain))
         goto end;
 
+#ifndef OPENSSL_NO_CNSM
+    if (cnsm_flag && SSL_CTX_use_certificate_file(ctx, key_enc_file, SSL_FILETYPE_PEM) <= 0)
+    {
```
Comment (paulidale) [inline apps/s_client.c]: Better might be:
```
if (cnsm_flag) {
    if (SSL_CTX_use_certificate_file(...) <= 0
        || SSL_CTX_use_enc_PrivateKey_file(...) <= 0
        || !SSL_CTX_check_enc_private_key(ctx)) {
        ERR_print_errors(bio_err);
        goto end;
    }
}
```

File: apps/s_server.c [hunk 19]
```diff
@@ -1654,6 +1680,10 @@ int s_server_main(int argc, char *argv[])
 
     if (s_key_file == NULL)
         s_key_file = s_cert_file;
+#ifndef OPENSSL_NO_CNSM
+    if (enc_flag!= 2 && s_key_enc_file == NULL)
```
Comment (paulidale) [inline apps/s_server.c]: Space before `!=`

File: apps/s_server.c [hunk 20]
```diff
@@ -2017,6 +2046,20 @@ int s_server_main(int argc, char *argv[])
         ERR_print_errors(bio_err);
         goto end;
     }
+	
+#ifndef OPENSSL_NO_CNSM
+    if (enc_flag && SSL_CTX_use_certificate_file(ctx, s_cert_enc_file, SSL_FILETYPE_PEM) <= 0)
```
Comment (paulidale) [inline apps/s_server.c]: Squash these like the above.

File: crypto/ec/ec_pmeth.c [hunk 21]
```diff
@@ -36,8 +36,37 @@ typedef struct {
     size_t kdf_ukmlen;
     /* KDF output length */
     size_t kdf_outlen;
+
+    #ifndef OPENSSL_NO_CNSM
```
Comment (paulidale) [inline crypto/ec/ec_pmeth.c]: No indentation on the _#_.

File: crypto/ec/ec_pmeth.c [hunk 22]
```diff
@@ -36,8 +36,37 @@ typedef struct {
...
+    int peerid_len;
+    /* self uid length */
+    int selfid_len;
+    /* peer ephemeral public key */
+    EC_KEY *peer_ecdhe_key;
+    /* self ephemeral key */
+    EC_KEY *self_ecdhe_key;
+    /* sm2/ecc encrypt out format, 0 for ASN1 */
+    int encdata_format;
+    #endif
```
Comment (paulidale) [inline crypto/ec/ec_pmeth.c]: No indentation for the _#_.

File: crypto/ec/ec_pmeth.c (hunk 10 above)
Comment (jntass) [reply crypto/ec/ec_pmeth.c]: Yes, it's a mistake indeed. Honestly, we don't test configured with no-cnsm  ever. I think i should check all of those macro define.

File: crypto/evp/m_sigver.c [hunk 23]
```diff
@@ -75,6 +81,21 @@ static int do_sigver_init(EVP_MD_CTX *ctx, EVP_PKEY_CTX **pctx,
         return 1;
     if (!EVP_DigestInit_ex(ctx, type, e))
         return 0;
+        
+#ifndef OPENSSL_NO_CNSM
+    if (ctx->pctx->pkey->type == EVP_PKEY_EC)
+    {
```
Comment (paulidale) [inline crypto/evp/m_sigver.c]: _{_ on previous line.

File: crypto/evp/pmeth_lib.c [hunk 24]
```diff
@@ -250,6 +250,13 @@ EVP_PKEY_CTX *EVP_PKEY_CTX_new_id(int id, ENGINE *e)
     return int_ctx_new(NULL, e, id);
 }
 
+#ifndef OPENSSL_NO_CNSM
+EVP_PKEY_CTX *EVP_PKEY_CTX_new_pkey_id(EVP_PKEY *pkey, int id, ENGINE *e)
```
Comment (paulidale) [inline crypto/evp/pmeth_lib.c]: Is it necessary to export the `int_ctx_new` function?
That's effectively what is being done here.

If it **must** be exported, renaming the original function is a better way.
I'm suspicious about it needing to be though.

File: crypto/sm2/sm2_crypt.c (hunk 11 above)
Comment (paulidale) [reply crypto/sm2/sm2_crypt.c]: Agreed.

File: crypto/sm2/sm2_crypt.c [hunk 25]
```diff
@@ -391,3 +391,350 @@ int sm2_decrypt(const EC_KEY *key,
...
+{
+    EVP_MD_CTX *mctx = NULL;
+    unsigned int counter;
+    unsigned char ctr[4];
+    size_t mdlen;
+    int retval = 0;
+
+    if (!out || !outlen) return retval;
```
Comment (paulidale) [inline crypto/sm2/sm2_crypt.c]: Line break after the `)`.

File: crypto/sm2/sm2_crypt.c [hunk 26]
```diff
@@ -391,3 +391,350 @@ int sm2_decrypt(const EC_KEY *key,
...
+        ctr[1] = (unsigned char)((counter >> 16) & 0xFF);
+        ctr[2] = (unsigned char)((counter >> 8) & 0xFF);
+        ctr[3] = (unsigned char)(counter & 0xFF);
+        if (!EVP_DigestUpdate(mctx, Z, Zlen))
+            goto err;
+        if (!EVP_DigestUpdate(mctx, ctr, sizeof(ctr)))
+            goto err;
+        if (!EVP_DigestUpdate(mctx, SharedInfo, SharedInfolen))
+            goto err;
+        if (!EVP_DigestFinal(mctx, dgst, NULL))
```
Comment (paulidale) [inline crypto/sm2/sm2_crypt.c]: Consider squashing these into one _if_ statement:
```
if (!EVP_DigestUpdate(...)
    && !EVP_DigestUpdate(...)
    && !EVP_DigestUpdate(...)
    && ! EVP_DigestFinal(...))
    goto err;
```

File: crypto/sm2/sm2_crypt.c [hunk 27]
```diff
@@ -391,3 +391,350 @@ int sm2_decrypt(const EC_KEY *key,
...
+
+    ctx = BN_CTX_new();
+    Xs = BN_new();
+    Xp = BN_new();
+    h = BN_new();
+    t = BN_new();
+    two_power_w = BN_new();
+    order = BN_new();
+
+    if (!Xs || !Xp || !h || !t || !two_power_w || !order)
```
Comment (paulidale) [inline crypto/sm2/sm2_crypt.c]: Only the last BN from the context needs to be checked.  I.e. `if (order == NULL)`

File: crypto/sm2/sm2_crypt.c [hunk 28]
```diff
@@ -391,3 +391,350 @@ int sm2_decrypt(const EC_KEY *key,
...
+    /*Sixth: Caculate Key -- Need Xuorv, Yuorv, Zc, Zs, klen*/
+    {
+        /*
+        size_t buflen, len;
+        unsigned char *buf = NULL;
+        */
+        size_t elemet_len, idx;
+
+        elemet_len = (size_t)((EC_GROUP_get_degree(group) + 7) / 8);
+        buflen = elemet_len * 2 + 32 * 2 + 1;    /*add 1 byte tag*/
```
Comment (paulidale) [inline crypto/sm2/sm2_crypt.c]: Where does the _32_ come from?  A comment is advisable.

File: crypto/sm2/sm2_crypt.c [hunk 29]
```diff
@@ -391,3 +391,350 @@ int sm2_decrypt(const EC_KEY *key,
...
+        buf = (unsigned char *)OPENSSL_malloc(buflen + 10);
```
Comment (paulidale) [inline crypto/sm2/sm2_crypt.c]: Likewise, why the magic _10_?

File: crypto/sm2/sm2_crypt.c [hunk 30]
```diff
@@ -391,3 +391,350 @@ int sm2_decrypt(const EC_KEY *key,
...
+        {
+            SM2err(SM2_F_SM2KAP_COMPUTE_KEY, ERR_R_INTERNAL_ERROR);
+            goto err;
+        }
+    }
+
+    ret = outlen;
+
+err:
+    if (Xs) BN_free(Xs);
```
Comment (paulidale) [inline crypto/sm2/sm2_crypt.c]: Just call `BN_free(Xs)`, NULL is fine as an argument.

File: ssl/ssl_rsa.c [hunk 31]
```diff
@@ -334,6 +343,94 @@ static int ssl_set_cert(CERT *c, X509 *x)
...
+		            ERR_clear_error();
+		        }
+		    }
+
+		    X509_free(c->pkeys[i].x509);
+		    X509_up_ref(x);
+		    c->pkeys[i].x509 = x;
+		    c->key = &(c->pkeys[i]);
+        }
+        if (x && (X509_get_extension_flags(x)& EXFLAG_KUSAGE) && (X509_get_key_usage(x) & (X509v3_KU_KEY_ENCIPHERMENT | X509v3_KU_DATA_ENCIPHERMENT | X509v3_KU_KEY_AGREEMENT))){
```
Comment (paulidale) [inline ssl/ssl_rsa.c]: This line **has** to be too long...

File: Configurations/unix-Makefile.tmpl (hunk 9 above)
Comment (levitte) [reply Configurations/unix-Makefile.tmpl]: You do misunderstand the usage. We have decided that for any new algorithm, the EVP API will be the only way to access it. That is even more important with the upcoming 3.0 release. That's the reason you will not find any public header file for the sm algos.

**Comment by t-j-h** [comment]
What implementations has this implementation been tested against for interoperability?

File: include/openssl/tls1.h (hunk 7 above)
Comment (jntass) [reply include/openssl/tls1.h]: After our discussion, we feel we should close this pull request. But we still want to thank you for helping us make some improvements and fix some bugs.

**Comment by jntass** [comment]
Tls1.0 and tls1.1 will be discarded , and our pull request is based on these version. As discussed in the above, we shall close this pull request. If you still want to get the tassl, please view our github: https://github.com/jntass/TASSL-1.1.1b

**Comment by zhaozg** [comment]
@jntass Do you plan to open a pr based on master again? If you can, please dont giveup, I have another pr base on v1.0.2, maybe I can do something for that

**Comment by paulidale** [comment]
A PR for v1.0.2 would need to be a security fix to be accepted and this doesn't seem like one.
A PR for master would definitely be considered for inclusion.

**Comment by jntass** [comment]
> @jntass Do you plan to open a pr based on master again? If you can, please dont giveup, I have another pr base on v1.0.2, maybe I can do something for that


As discussed above, tassl is base on tls1.1 (https://tools.ietf.org/html/draft-ietf-tls-oldversions-deprecate). There's lots of security problems known by public in 2019. So we don't think it's a good idea pull this PR to the official openssl. But we still welcome you can do something for tassl.

File: crypto/sm2/sm2_crypt.c (hunk 25 above)
Comment (jntass) [reply crypto/sm2/sm2_crypt.c]: Thanks for your carefully review.

File: crypto/sm2/sm2_crypt.c (hunk 26 above)
Comment (jntass) [reply crypto/sm2/sm2_crypt.c]: Maybe it should be '||' instead of '&&'. Because i want goto err if there is anyone is true.
